STM32/SPI/Src/ADS7953.c

```diff
@@ -0,0 +1,520 @@
+static void bufferHalfFullCallback(DMA_HandleTypeDef *hdma);
+static void bufferFullCallback(DMA_HandleTypeDef *hdma);
+static void initDMA(ADS7953Device_t *dev);
+static void initTimer(ADS7953Device_t *dev);
+
+/***************************************************************************************************
+** PRIVATE OBJECTS
+***************************************************************************************************/
+
+/* 
+This array stores the adresses of all the ADS7953 initialized
+It is needed in bufferHalfFullCallback() and bufferFullCallback()
+as the device pointer cannot be passed directly
+*/
+static ADS7953Device_t *listOfDevices[MAX_DEVICES_NO] = {NULL};
+
+/***************************************************************************************************
+** PRIVATE FUNCTION DEFINITIONS
+***************************************************************************************************/
+
+/*!
+ * @brief   Create part of message to setup the last ADC channel
+ * @param   noOfInputs Number of ADC channels used
+ * @return  Part of message
+*/
+static uint16_t lastInputProgramming(uint16_t noOfInputs)
+{
+    return (((noOfInputs - 1) & 0x0FU) << 6);
+}
+
+/*!
+ * @brief   Extraction of channel address from SPI message
+ * @param   message SPI message received
+ * @return  ADC channel address
+*/
+static uint16_t getChannelAddress(uint16_t message)
+{
+    return ((message & 0xF000U) >> 12);
+}
+
+/*!
+ * @brief   Extraction of ADC value from SPI message
+ * @param   message SPI message received
+ * @return  12-bits ADC value
+*/
+static uint16_t getConversionResult(uint16_t message)
+{
+    return (message & 0x0FFFU);
+}
+
+/*!
+ * @brief   Creation of Mode Control Register Message
+ * @return  SPI message
+*/
+static uint16_t createModeCtrllRegMessage()
+{
+    static const uint16_t AUTO_2_MODE_SELECT = 0x3000;
+    static const uint16_t ENABLE_PROGRAMMING = 0x0800;
+    static const uint16_t CHANNEL_COUNTER_RESET = 0x0400;
+    static const uint16_t DOUBLE_V_REF_SELECT = 0x0040;
+
+    return (AUTO_2_MODE_SELECT | ENABLE_PROGRAMMING | CHANNEL_COUNTER_RESET | DOUBLE_V_REF_SELECT);
+}
+
+/*!
+ * @brief   Creation of Program Register Message
+ * @param   noOfChannels Number of ADC channels used
+ * @return  SPI message
+*/
+static uint16_t createProgramRegMessage(uint16_t noOfChannels)
+{
+    uint16_t AUTO_2_PROGRAM_REGISTER = 0x9000U;
+
+    return (AUTO_2_PROGRAM_REGISTER | lastInputProgramming(noOfChannels));
+}
+
+/*!
+ * @brief   Sends 2 configuration messages to the ADS7953 selected
+ * @param   dev Pointer to the ADC structure
+*/
+static int setRegisters(ADS7953Device_t *dev)
+{
+    // Messages
+    ADS7953Message_t modeControlRegisterMessage = { .fullMessage = createModeCtrllRegMessage() };
+    ADS7953Message_t programRegisterMessage = { .fullMessage = createProgramRegMessage(dev->noOfChannels) };
+
+    HAL_Delay(1);                   // Necessary to wait until the NSS pin is high (pull-up)
```

Comment:
I doesn't look like this delay is necessary :thinking: I see how it is necessary between messages, but I don't think it help you here.